File: foundation/foundation-core-java/src/test/java/com/boozallen/aissemble/core/policy/configuration/util/PolicyTestUtil.java
package com.boozallen.aissemble.core.policy.configuration.util;

/*-
 * #%L
 * Policy-Based Configuration::Policy Manager
 * %%
 * Copyright (C) 2021 Booz Allen
 * %%
 * This software package is licensed under the Booz Allen Public License. All Rights Reserved.
 * #L%
 */

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.aeonbits.owner.KrauseningConfigFactory;
import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang3.RandomStringUtils;
import org.apache.commons.lang3.RandomUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.boozallen.aissemble.core.policy.configuration.PolicyConfiguration;
import com.boozallen.aissemble.core.policy.configuration.policy.Policy;
import com.boozallen.aissemble.core.policy.configuration.policy.Target;
import com.boozallen.aissemble.core.policy.configuration.policy.json.PolicyInput;
import com.boozallen.aissemble.core.policy.configuration.policy.json.rule.PolicyRuleInput;
import com.boozallen.aissemble.core.policy.configuration.policymanager.AbstractPolicyManager;
import com.fasterxml.jackson.databind.ObjectMapper;

public class PolicyTestUtil {

    private static final String[] availableAlgorithms = new String[] { "a.fake.package.and.classname",
            "com.different.company.algorithm.CustomAlgorithm" };

    private static final Logger logger = LoggerFactory.getLogger(PolicyTestUtil.class);

    private static ObjectMapper mapper = new ObjectMapper();
    
    private static PolicyConfiguration configuration = KrauseningConfigFactory.create(PolicyConfiguration.class);

    public static String getRandomAlgorithm() {
        int index = RandomUtils.nextInt(0, availableAlgorithms.length);
        return availableAlgorithms[index];
    }

    public static void writePolicyToDefaultLocation(List<PolicyInput> policies, String fileName) {
        writePoliciesToFile(policies, configuration.getPoliciesLocation(), fileName);
    }

    public static void readPoliciesFromDefaultLocation(AbstractPolicyManager policyManager) {

        // Load the files from the configured path
        Map<String, Policy> policies = policyManager.getPolicies();
        policies.clear();
        policyManager.loadPolicyConfigurations(configuration.getPoliciesLocation());
    }

    /**
     * Writes a policy to a file in the directory passed in.
     * 
     * @param policies
     *            policies to be written out
     * @param directory
     *            the directory to write to
     * @param file
     *            policy filename
     */
    public static void writePoliciesToFile(List<PolicyInput> policies, String directory, String file) {

        // Create the parent directory if need be
        File parentDirectory = new File(directory);
        if (!parentDirectory.exists()) {
            parentDirectory.mkdirs();
        }

        // Write the file
        String jsonFile = FilenameUtils.concat(directory, file);
        FileWriter writer = null;
        try {
            writer = new FileWriter(jsonFile);
            String json = mapper.writeValueAsString(policies);
            writer.write(json);
        } catch (Exception e) {
            logger.error("Error writing a file to: {}. Error was: {}", jsonFile, e);
        } finally {
            if (writer != null) {
                try {
                    writer.close();
                } catch (IOException e) {
                    logger.error("File writer didn't close but honestly, this test is almost over.", e);
                }
            }
        }
    }

    /**
     * Convenience method for wrapping a policy in a list and then writing it to
     * file.
     * 
     * @param policy
     * @param directory
     * @param file
     */
    public static void writePolicyToFile(PolicyInput policy, String directory, String file) {
        writePoliciesToFile(Arrays.asList(policy), directory, file);
    }

    public static PolicyInput getRandomPolicy() {
        PolicyInput policyInput = new PolicyInput(RandomStringUtils.randomAlphabetic(10));
        policyInput.setTarget(getRandomTarget());
        List<PolicyRuleInput> rules = getRandomRules(RandomUtils.nextInt(2, 6));
        policyInput.setRules(rules);
        return policyInput;
    }

    /**
     * Method that creates some random rules. Switches between the different
     * json configs for fun and variety.
     * 
     * @param number
     * @return randomRules
     */
    public static List<PolicyRuleInput> getRandomRules(int number) {
        List<PolicyRuleInput> rules = new ArrayList<PolicyRuleInput>();
        String algorithm = getRandomAlgorithm();
        for (int i = 0; i < number; i++) {

            // Get a good mix of rule configurations written out. Just choose
            // randomly.
            int random = RandomUtils.nextInt(1, 5);
            PolicyRuleInput randomRule = new PolicyRuleInput(algorithm);
            switch (random) {
            case 1:
                randomRule = new PolicyRuleInput(algorithm, getRandomConfigs(), null);
                break;
            case 2:
                randomRule = new PolicyRuleInput(algorithm, null, getRandomConfigs());
                break;
            case 3:
                randomRule = new PolicyRuleInput(algorithm, getRandomConfigs(), getRandomConfigs());
                break;
            }
            rules.add(randomRule);

        }
        return rules;
    }

    private static Target getRandomTarget() {
        Target target = new Target("http://" + RandomStringUtils.randomAlphabetic(5) + ".com",
                RandomStringUtils.randomAlphabetic(5));
        return target;
    }

    private static Map<String, Object> getRandomConfigs() {
        Map<String, Object> map = new HashMap<String, Object>();
        map.put(RandomStringUtils.randomAlphabetic(5), RandomStringUtils.randomAlphabetic(3));
        map.put(RandomStringUtils.randomAlphabetic(3), RandomUtils.nextInt(1, 100));
        return map;
    }

}

File: foundation/aissemble-foundation-core-python/tests/features/steps/policy_test_util.py
###
# #%L
# Policy-Based Configuration::Policy Manager (Python)
# %%
# Copyright (C) 2021 Booz Allen
# %%
# This software package is licensed under the Booz Allen Public License. All Rights Reserved.
# #L%
###
from typing import Any, Dict, List
from policy_manager import AbstractPolicyManager
from policy_manager.policy import Target
from policy_manager.policy.json import (
    PolicyInput,
    PolicyRuleInput,
)
from policy_manager.configuration import PolicyConfiguration
from krausening.logging import LogManager
import random, string, json, os
from os import environ

environ["KRAUSENING_BASE"] = "./tests/resources/krausening/base"


class PolicyTestUtil:
    availableAlgorithms = [
        "a.fake.package.and.classname",
        "com.different.company.algorithm.CustomAlgorithm",
    ]
    logger = LogManager.get_instance().get_logger("PolicyTestUtil")
    configuration = PolicyConfiguration()

    @staticmethod
    def getRandomAlgorithm() -> str:
        return random.choice(PolicyTestUtil.availableAlgorithms)

    @staticmethod
    def writePolicyToDefaultLocation(
        policies: List[PolicyInput], fileName: str
    ) -> None:
        PolicyTestUtil.writePoliciesToFile(
            policies, PolicyTestUtil.configuration.policiesLocation(), fileName
        )

    @staticmethod
    def readPoliciesFromDefaultLocation(policyManager: AbstractPolicyManager) -> None:
        policies = policyManager.policies
        policies.clear()
        policyManager.loadPolicyConfigurations(
            PolicyTestUtil.configuration.policiesLocation()
        )

    @staticmethod
    def writePoliciesToFile(
        policies: List[PolicyInput], directory: str, file: str
    ) -> None:
        filePath = os.path.join(directory, file)

        # Create the parent directory if need be
        os.makedirs(os.path.dirname(filePath), exist_ok=True)

        # Write the file
        with open(filePath, "w") as jsonFile:
            content = [policy.dict(exclude_none=True) for policy in policies]
            json.dump(content, jsonFile)

    @staticmethod
    def writePolicyToFile(policy: PolicyInput, directory: str, file: str) -> None:
        PolicyTestUtil.writePoliciesToFile([policy], directory, file)

    @staticmethod
    def getRandomPolicy() -> PolicyInput:
        identifier = PolicyTestUtil.getRandomString(10)
        target = PolicyTestUtil.getRandomTarget()
        rules = PolicyTestUtil.getRandomRules(PolicyTestUtil.getRandomNumber(2, 6))
        return PolicyInput(identifier=identifier, target=target, rules=rules)

    @staticmethod
    def getRandomRules(number: int) -> List[PolicyRuleInput]:
        rules = []
        algorithm = PolicyTestUtil.getRandomAlgorithm()

        for index in range(number):
            # Get a good mix of rule configurations written out. Just choose randomly.
            random = PolicyTestUtil.getRandomNumber(1, 5)
            randomRule = PolicyRuleInput(className=algorithm)
            if random == 1:
                randomRule = PolicyRuleInput(
                    className=algorithm,
                    configurations=PolicyTestUtil.getRandomConfigs(),
                )
            if random == 2:
                randomRule = PolicyRuleInput(
                    className=algorithm,
                    targetConfigurations=PolicyTestUtil.getRandomConfigs(),
                )
            if random == 3:
                randomRule = PolicyRuleInput(
                    className=algorithm,
                    configurations=PolicyTestUtil.getRandomConfigs(),
                    targetConfigurations=PolicyTestUtil.getRandomConfigs(),
                )

            rules.append(randomRule)

        return rules

    @staticmethod
    def getRandomTarget() -> Target:
        retrieveUrl = "http://" + PolicyTestUtil.getRandomString(5) + ".com"
        type = PolicyTestUtil.getRandomString(5)
        return Target(retrieve_url=retrieveUrl, type=type)

    @staticmethod
    def getRandomConfigs() -> Dict[str, Any]:
        configs = dict()
        configs[PolicyTestUtil.getRandomString(5)] = PolicyTestUtil.getRandomString(3)
        configs[PolicyTestUtil.getRandomString(3)] = PolicyTestUtil.getRandomNumber(
            1, 100
        )

        return configs

    @staticmethod
    def getRandomString(length: int) -> str:
        return "".join(random.choice(string.ascii_letters) for i in range(length))

    @staticmethod
    def getRandomNumber(min: int, max: int) -> int:
        return random.randint(min, max)
